ext/nnstreamer/tensor_filter/tensor_filter_llamacpp.cc

```diff
@@ -214,26 +249,59 @@ TensorFilterLlamaCpp::invoke (const GstTensorMemory *input, GstTensorMemory *out
 }
 
 /**
- * @brief Invoke llamacpp using input tensors
+ * @brief Create output tensor
+ *
+ * The memory allocated for the output tensor data using strndup
+ * must be freed by the caller of this function to avoid memory leaks.
+ */
+GstTensorMemory *
+TensorFilterLlamaCpp::createOutputTensor (
+    const char *buf, int n, GstTensorFilterProperties *prop)
+{
+  if (buf == nullptr || n <= 0 || prop == nullptr) {
+    ml_loge ("Invalid arguments passed to createOutputTensor");
+    return nullptr;
+  }
+
+  GstTensorMemory *output = g_new0 (GstTensorMemory, 1);
+  if (!output) {
+    ml_loge ("createOutputTensor: Allocation failed");
+    return nullptr;
+  }
+
+  output[0].data = g_strndup (buf, n);
+  output[0].size = n; /* Stores string length, excluding NULL appended to the end of the string */
+  if (output[0].data == nullptr) {
+    ml_loge ("createOutputTensor: strndup failed");
+    g_free (output);
+    return nullptr;
+  }
+
+  gst_tensors_info_free (&prop->output_meta);
+  prop->output_meta.num_tensors = 1;
+  prop->output_meta.format = _NNS_TENSOR_FORMAT_FLEXIBLE;
+  prop->output_meta.info[0].type = _NNS_UINT8;
+  prop->output_meta.info[0].dimension[0] = output[0].size;
+
+  return output;
+}
+
+/**
+ * @brief Generates output tokens based on the provided input prompt using the
+ * llamacpp and dispatches the generated tokens to tensor_filter asynchronously.
+ * nnstreamer frees GstTensorMemory output after use. Do not free it.
  */
 void
-TensorFilterLlamaCpp::invoke_dynamic (GstTensorFilterProperties *prop,
-    const GstTensorMemory *input, GstTensorMemory *output)
+TensorFilterLlamaCpp::generateTokens (GstTensorFilterProperties *prop,
+    std::string &&prompt, GstTensorMemory *output)
 {
   const llama_vocab *vocab = llama_model_get_vocab (model);
-  llama_context *ctx = nullptr;
```

Comment:
plz consider ctx, newly allocating local value and releasing in generateTokens() method.

Reply:
Thank you. I've updated it.